File: media_manager_app/lib/src/core/models/studio.dart
import 'package:json_annotation/json_annotation.dart';

part 'studio.g.dart';

/// 制作商实体
@JsonSerializable(fieldRename: FieldRename.snake)
class Studio {
  final String id;
  final String name;
  final String? logoUrl;
  final String? description;
  final int mediaCount;
  final DateTime createdAt;
  final DateTime updatedAt;

  const Studio({
    required this.id,
    required this.name,
    this.logoUrl,
    this.description,
    required this.mediaCount,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Studio.fromJson(Map<String, dynamic> json) => _$StudioFromJson(json);
  Map<String, dynamic> toJson() => _$StudioToJson(this);

  Studio copyWith({
    String? id,
    String? name,
    String? logoUrl,
    String? description,
    int? mediaCount,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Studio(
      id: id ?? this.id,
      name: name ?? this.name,
      logoUrl: logoUrl ?? this.logoUrl,
      description: description ?? this.description,
      mediaCount: mediaCount ?? this.mediaCount,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }
}

/// 系列实体
@JsonSerializable(fieldRename: FieldRename.snake)
class Series {
  final String id;
  final String name;
  final String? studioId;
  final String? description;
  final String? coverUrl;
  final int mediaCount;
  final DateTime createdAt;
  final DateTime updatedAt;

  const Series({
    required this.id,
    required this.name,
    this.studioId,
    this.description,
    this.coverUrl,
    required this.mediaCount,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Series.fromJson(Map<String, dynamic> json) => _$SeriesFromJson(json);
  Map<String, dynamic> toJson() => _$SeriesToJson(this);

  Series copyWith({
    String? id,
    String? name,
    String? studioId,
    String? description,
    String? coverUrl,
    int? mediaCount,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Series(
      id: id ?? this.id,
      name: name ?? this.name,
      studioId: studioId ?? this.studioId,
      description: description ?? this.description,
      coverUrl: coverUrl ?? this.coverUrl,
      mediaCount: mediaCount ?? this.mediaCount,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }
}

/// 带制作商信息的系列
@JsonSerializable(fieldRename: FieldRename.snake)
class SeriesWithStudio {
  final String id;
  final String name;
  final String? studioId;
  final String? description;
  final String? coverUrl;
  final int mediaCount;
  final DateTime createdAt;
  final DateTime updatedAt;
  final String? studioName;

  const SeriesWithStudio({
    required this.id,
    required this.name,
    this.studioId,
    this.description,
    this.coverUrl,
    required this.mediaCount,
    required this.createdAt,
    required this.updatedAt,
    this.studioName,
  });

  factory SeriesWithStudio.fromJson(Map<String, dynamic> json) =>
      _$SeriesWithStudioFromJson(json);
  Map<String, dynamic> toJson() => _$SeriesWithStudioToJson(this);

  Series toSeries() => Series(
        id: id,
        name: name,
        studioId: studioId,
        description: description,
        coverUrl: coverUrl,
        mediaCount: mediaCount,
        createdAt: createdAt,
        updatedAt: updatedAt,
      );
}

/// 带系列列表的制作商
@JsonSerializable(fieldRename: FieldRename.snake)
class StudioWithSeries {
  final String id;
  final String name;
  final String? logoUrl;
  final String? description;
  final int mediaCount;
  final DateTime createdAt;
  final DateTime updatedAt;
  final List<Series> seriesList;

  const StudioWithSeries({
    required this.id,
    required this.name,
    this.logoUrl,
    this.description,
    required this.mediaCount,
    required this.createdAt,
    required this.updatedAt,
    required this.seriesList,
  });

  factory StudioWithSeries.fromJson(Map<String, dynamic> json) =>
      _$StudioWithSeriesFromJson(json);
  Map<String, dynamic> toJson() => _$StudioWithSeriesToJson(this);

  Studio toStudio() => Studio(
        id: id,
        name: name,
        logoUrl: logoUrl,
        description: description,
        mediaCount: mediaCount,
        createdAt: createdAt,
        updatedAt: updatedAt,
      );
}

/// 制作商列表响应
@JsonSerializable(fieldRename: FieldRename.snake)
class StudioListResponse {
  final List<StudioWithSeries> studios;
  final int total;

  const StudioListResponse({
    required this.studios,
    required this.total,
  });

  factory StudioListResponse.fromJson(Map<String, dynamic> json) =>
      _$StudioListResponseFromJson(json);
  Map<String, dynamic> toJson() => _$StudioListResponseToJson(this);
}

/// 系列列表响应
@JsonSerializable(fieldRename: FieldRename.snake)
class SeriesListResponse {
  final List<SeriesWithStudio> series;
  final int total;

  const SeriesListResponse({
    required this.series,
    required this.total,
  });

  factory SeriesListResponse.fromJson(Map<String, dynamic> json) =>
      _$SeriesListResponseFromJson(json);
  Map<String, dynamic> toJson() => _$SeriesListResponseToJson(this);
}

// ============ Request DTOs ============

/// 创建制作商请求
class CreateStudioRequest {
  final String name;
  final String? logoUrl;
  final String? description;

  const CreateStudioRequest({
    required this.name,
    this.logoUrl,
    this.description,
  });

  Map<String, dynamic> toJson() => {
        'name': name,
        if (logoUrl != null) 'logo_url': logoUrl,
        if (description != null) 'description': description,
      };
}

/// 更新制作商请求
class UpdateStudioRequest {
  final String? name;
  final String? logoUrl;
  final String? description;

  const UpdateStudioRequest({
    this.name,
    this.logoUrl,
    this.description,
  });

  Map<String, dynamic> toJson() => {
        if (name != null) 'name': name,
        if (logoUrl != null) 'logo_url': logoUrl,
        if (description != null) 'description': description,
      };
}

/// 创建系列请求
class CreateSeriesRequest {
  final String name;
  final String? studioId;
  final String? studioName;
  final String? description;
  final String? coverUrl;

  const CreateSeriesRequest({
    required this.name,
    this.studioId,
    this.studioName,
    this.description,
    this.coverUrl,
  });

  Map<String, dynamic> toJson() => {
        'name': name,
        if (studioId != null) 'studio_id': studioId,
        if (studioName != null) 'studio_name': studioName,
        if (description != null) 'description': description,
        if (coverUrl != null) 'cover_url': coverUrl,
      };
}

/// 更新系列请求
class UpdateSeriesRequest {
  final String? name;
  final String? studioId;
  final String? description;
  final String? coverUrl;

  const UpdateSeriesRequest({
    this.name,
    this.studioId,
    this.description,
    this.coverUrl,
  });

  Map<String, dynamic> toJson() => {
        if (name != null) 'name': name,
        if (studioId != null) 'studio_id': studioId,
        if (description != null) 'description': description,
        if (coverUrl != null) 'cover_url': coverUrl,
      };
}

File: media_manager_backend/src/models/studio.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use sqlx::FromRow;

/// 厂商/制作公司
#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct Studio {
    pub id: String,
    pub name: String,
    pub logo_url: Option<String>,
    pub description: Option<String>,
    pub media_count: i32,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

impl Studio {
    pub fn new(name: String) -> Self {
        let now = Utc::now();
        Self {
            id: uuid::Uuid::new_v4().to_string(),
            name,
            logo_url: None,
            description: None,
            media_count: 0,
            created_at: now,
            updated_at: now,
        }
    }
}

/// 系列
#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct Series {
    pub id: String,
    pub name: String,
    pub studio_id: Option<String>,
    pub description: Option<String>,
    pub cover_url: Option<String>,
    pub media_count: i32,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

impl Series {
    pub fn new(name: String, studio_id: Option<String>) -> Self {
        let now = Utc::now();
        Self {
            id: uuid::Uuid::new_v4().to_string(),
            name,
            studio_id,
            description: None,
            cover_url: None,
            media_count: 0,
            created_at: now,
            updated_at: now,
        }
    }
}

/// 带厂商信息的系列（用于API响应）
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SeriesWithStudio {
    #[serde(flatten)]
    pub series: Series,
    pub studio_name: Option<String>,
}

/// 带系列列表的厂商（用于API响应）
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StudioWithSeries {
    #[serde(flatten)]
    pub studio: Studio,
    pub series_list: Vec<Series>,
}

// ============ Request/Response DTOs ============

#[derive(Debug, Deserialize)]
pub struct CreateStudioRequest {
    pub name: String,
    pub logo_url: Option<String>,
    pub description: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct UpdateStudioRequest {
    pub name: Option<String>,
    pub logo_url: Option<String>,
    pub description: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct CreateSeriesRequest {
    pub name: String,
    pub studio_id: Option<String>,
    pub studio_name: Option<String>,  // 可通过厂商名称关联
    pub description: Option<String>,
    pub cover_url: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct UpdateSeriesRequest {
    pub name: Option<String>,
    pub studio_id: Option<String>,
    pub description: Option<String>,
    pub cover_url: Option<String>,
}

#[derive(Debug, Serialize)]
pub struct StudioListResponse {
    pub studios: Vec<StudioWithSeries>,
    pub total: i64,
}

#[derive(Debug, Serialize)]
pub struct SeriesListResponse {
    pub series: Vec<SeriesWithStudio>,
    pub total: i64,
}

/// 系列匹配结果（用于导入时的智能匹配）
#[derive(Debug, Clone, Serialize)]
pub struct SeriesMatchResult {
    pub series_id: String,
    pub series_name: String,
    pub studio_id: Option<String>,
    pub studio_name: Option<String>,
    pub match_type: SeriesMatchType,
}

#[derive(Debug, Clone, Serialize)]
pub enum SeriesMatchType {
    Exact,           // 精确匹配（系列名+厂商都匹配）
    UniqueByName,    // 系列名唯一，自动关联厂商
    Ambiguous,       // 系列名存在于多个厂商，需要确认
    NewSeries,       // 新系列
}
